internal/util/logging/pgx.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !ferretdb_no_postgresql
+
+package logging
+
+import (
+	"context"
+	"log/slog"
+	"runtime"
+	"slices"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/jackc/pgx/v5/tracelog"
+	"golang.org/x/exp/maps"
+
+	"github.com/FerretDB/FerretDB/internal/util/debugbuild"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+// pgxLogLevels maps pgx log levels to slog levels.
+var pgxLogLevels = map[tracelog.LogLevel]slog.Level{
+	// pgx's info level is used for queries, but we want to hide them by default
+	tracelog.LogLevelTrace: slog.LevelDebug - 2,
+	tracelog.LogLevelDebug: slog.LevelDebug,
+	tracelog.LogLevelInfo:  slog.LevelDebug + 2,
+
+	tracelog.LogLevelWarn:  slog.LevelWarn,
+	tracelog.LogLevelError: slog.LevelError,
+}
+
+// pgxLogger is a pgx's [tracelog.Logger] implementation that uses slog
+// with correct source code location.
+type pgxLogger struct {
+	l *slog.Logger
+}
+
+// NewPgxLogger creates a new [tracelog.Logger] that uses given [slog.Logger].
+func NewPgxLogger(l *slog.Logger) tracelog.Logger {
+	return &pgxLogger{l: l}
+}
+
+// Log implements [tracelog.Logger].
+func (sl *pgxLogger) Log(ctx context.Context, level tracelog.LogLevel, msg string, data map[string]any) {
+	now := time.Now()
+
+	l, ok := pgxLogLevels[level]
+	if !ok {
+		l = slog.LevelError + 1
+		sl.l.LogAttrs(
+			ctx, slog.LevelWarn, "Invalid pgx log level",
+			slog.Int64("level", int64(level)), slog.String("original_msg", msg),
+		)
+	}
+
+	if !sl.l.Enabled(ctx, l) {
+		return
+	}
+
+	// We can't call runtime.Caller(s) with some fixed skip value and then inspect a single frame
+	// because call Pgx calls Log method from different stack depths.
+
+	var pc uintptr
+	callers := make([]uintptr, 10)
+
+	if n := runtime.Callers(2, callers); n != 0 {
+		frames := runtime.CallersFrames(callers[:n])
+
+		for {
```

Comment:
One question I have here (and also in `FuncCall`) is how much overhead we bring by calling all these additional things, such as `frames.Next()`. As far as I understand, it doesn't bring a lot of overhead as we don't "calculate" any special things, they already exist, but still, it would be nice to run some benchmarks with and without these logs being turned on.

Reply:
I like when you answer your own questions :-)